src/main/scala/kamon/jdbc/instrumentation/StatementInstrumentation.scala

```diff
@@ -15,104 +15,146 @@
+    val customRecorder = target.asInstanceOf[HasStatementMetrics.Mixin].statementMetrics
+    val entityRecorder = if(customRecorder != null) customRecorder else JdbcExtension.defaultTracker
+    val latencyRecorder = statementType match {
+      case StatementTypes.Query           ⇒ entityRecorder.queries
+      case StatementTypes.Update          ⇒ entityRecorder.updates
+      case StatementTypes.Batch           ⇒ entityRecorder.batches
+      case StatementTypes.GenericExecute  ⇒ entityRecorder.genericExecute
     }
-  } getOrElse pjp.proceed()
 
-  def withTimeSpent[A](thunk: ⇒ A)(timeSpent: Long ⇒ Unit): A = {
-    val start = System.nanoTime()
-    try thunk finally timeSpent(System.nanoTime() - start)
+    trackExecution(pjp, sql, statementType, latencyRecorder, entityRecorder.inFlightStatements, entityRecorder.errors, entityRecorder.slowStatements)
   }
 
-  def withSegment[A](ctx: TraceContext, statement: String)(thunk: ⇒ A): A = {
-    val segmentName = JdbcExtension.generateJdbcSegmentName(statement)
-    val segment = ctx.startSegment(segmentName, SegmentCategory.Database, JdbcExtension.SegmentLibraryName)
-    try thunk finally segment.finish()
-  }
+  def trackExecution(pjp: ProceedingJoinPoint, sql: String, statementType: String, latencyHistogram: Histogram,
+    inFlightCounter: MinMaxCounter, errorCounter: Counter, slowStatementsCounter: Counter): Any = {
 
-  def recordRead(pjp: ProceedingJoinPoint, sql: String)(implicit statementRecorder: StatementsMetrics): Any = {
-    withTimeSpent(pjp.proceedWithErrorHandler(sql)) { timeSpent ⇒
-      statementRecorder.reads.record(timeSpent)
+    inFlightCounter.increment()
+    val startedAt = System.nanoTime()
 
-      val timeSpentInMillis = nanos.toMillis(timeSpent)
+    try {
+      if (Tracer.currentContext.nonEmpty && JdbcExtension.shouldGenerateSegments)
+        withSegment(Tracer.currentContext, statementType, sql)(pjp.proceed())
+      else
+        pjp.proceed()
+
+    } catch {
+      case NonFatal(throwable) ⇒ {
+        JdbcExtension.processSqlError(sql, throwable)
+        errorCounter.increment()
+        throw throwable
+      }
+    } finally {
+      val timeSpent = System.nanoTime() - startedAt
+      latencyHistogram.record(timeSpent)
+      inFlightCounter.decrement()
 
+      val timeSpentInMillis = nanos.toMillis(timeSpent)
       if (timeSpentInMillis >= JdbcExtension.slowQueryThreshold) {
-        statementRecorder.slows.increment()
+        slowStatementsCounter.increment()
         JdbcExtension.processSlowQuery(sql, timeSpentInMillis)
       }
     }
   }
 
-  def recordWrite(pjp: ProceedingJoinPoint, sql: String)(implicit statementRecorder: StatementsMetrics): Any = {
-    withTimeSpent(pjp.proceedWithErrorHandler(sql)) { timeSpent ⇒
-      statementRecorder.writes.record(timeSpent)
-    }
+  def withSegment[A](ctx: TraceContext, statementType: String, statement: String)(thunk: ⇒ A): A = {
```

Comment:
I think we need drive the case when the segment is finish with error and finish the segment with:`segment.finishWithError(Throwable)`. @ivantopo WDYT?